src/status.rs

```diff
@@ -247,44 +365,95 @@ pub fn show_build_status(
+			None
+		} else if simulated_time < now {
+			// Should have been done already, estimate it will be done immediately
+			Some(Duration::from_millis(0))
 		} else {
-			let remaining_duration = simulated_time - now;
-			let eta = TimeDuration::from_std(remaining_duration)
-				.map(|duration| chrono::Local::now() + duration);
-			match eta {
-				Ok(eta) => {
-					let is_soon = eta.date() == chrono::Local::now().date()
-						&& remaining_duration < Duration::from_secs(8 * 3600);
-					let timeformat = if is_soon {
-						"%H:%M:%S"
+			Some(simulated_time - now)
+		};
+
+		let (percentage, text) = match remaining_duration {
+			None => (0., format!("??% (estimating time)")),
+			Some(remaining_duration) => {
+				let percentage = (100 * as_millis(current_duration)) as f32
+					/ as_millis(current_duration + remaining_duration) as f32;
+
+				// Every 5 seconds, switch between showing ETA and remaining duration
+				let show_eta = (current_duration.as_secs() % 10) > 5;
+
+				(
+					percentage,
+					if show_eta {
+						let eta = TimeDuration::from_std(remaining_duration)
+							.map(|duration| chrono::Local::now() + duration)
+							.unwrap();
+						let is_soon = eta.date() == chrono::Local::now().date()
+							&& remaining_duration < Duration::from_secs(8 * 3600);
+						let timeformat = if is_soon {
+							"%H:%M:%S"
+						} else {
+							"%Y-%m-%d %H:%M:%S"
+						};
+						format!(
+							"{}% (ETA {})",
+							percentage.ceil() as u8,
+							eta.format(timeformat)
+						)
 					} else {
-						"%Y-%m-%d %H:%M:%S"
-					};
-					println!(
-						"Building for {}, remaining time for build is {} (ETA {})\x1b[K\x1b[m",
-						MinSec::since(start_time),
-						MinSec::from_duration(remaining_duration),
-						eta.format(timeformat)
-					);
-				}
-				_ => println!(
-					"Building for {}, remaining time is infinite...\x1b[K\x1b[m",
-					MinSec::since(start_time)
-				),
+						format!(
+							"{}% ({} remaining)",
+							percentage.ceil() as u8,
+							MinSec::from_duration(remaining_duration)
+						)
+					},
+				)
 			}
-		}
+		};
+
+		let progress = match progress_format {
+			ProgressFormat::None => "".to_owned(),
+			ProgressFormat::Text => format!(
+				"[Building for {}, {}]\x1b[K\x1b[m\n",
+				MinSec::since(start_time),
+				text
+			),
+			ProgressFormat::ASCIIBar | ProgressFormat::HighResBar => format!(
+				"[{}]\x1b[K\x1b[m\n",
+				render_progress_bar(
+					terminal_width() - 3,
+					percentage,
+					&text,
+					progress_format == ProgressFormat::HighResBar
+				)
+			),
+		};
+
+		print!("{}", progress);
 
 		if build_is_done {
 			break;
 		}
 
-		print!("\x1b[{}A", buildstate.workers.len() + 1);
+		print!(
+			"\x1b[{}A",
+			buildstate.workers.len() + progress.lines().count()
+		);
 		lock = status.inner.lock().unwrap();
 	}
 	println!("\x1b[32;1mFinished.\x1b[m");
 }
+
+fn terminal_width() -> usize {
+	if let Some((w, _)) = term_size::dimensions() {
```

Comment:
This will do an `iotcl` syscall every time. In C, one would only check on the `WINCH` signal, but not sure how you can do that in Rust.

Reply:
I expect the `term_size` crate to implement caching options like that, eventually.

Reply:
And somehow expect it to know when to invalidate that cache? The crate hasn't been updated since 2017, so it doesn't look like it's still going to get any new features.

Anyway, it's not too bad, since every time we do this `ioctl`, we also do a few `write`s, so we're not blowing up the amount of syscalls that much.

Later when we handle signals like `SIGINT` for aborting, we should also handle `SIGWINCH` and only do this `ioctl` when needed.